Sources/AsyncHTTPClient/FileDownloadDelegate.swift

```diff
@@ -0,0 +1,88 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2020 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+/// Handles a streaming download to a given file path, allowing headers and progress to be reported.
+public final class FileDownloadDelegate: HTTPClientResponseDelegate {
+    /// The response type for this delegate: the total count of bytes as reported by the response
+    /// "Content-Length" header (if available) and the count of bytes downloaded.
+    public typealias Response = (totalBytes: Int?, receivedBytes: Int)
+
+    private var totalBytes: Int?
+    private var receivedBytes = 0
+
+    private let handle: NIOFileHandle
+    private let io: NonBlockingFileIO
+    private let reportHeaders: ((HTTPHeaders) -> Void)?
+    private let reportProgress: ((_ totalBytes: Int?, _ receivedBytes: Int) -> Void)?
+
+    private var writeFuture: EventLoopFuture<Void>?
+
+    /// Initializes a new file download delegate.
+    /// - parameters:
+    ///     - path: Path to a file you'd like to write the download to.
+    ///     - pool: A thread pool to use for asynchronous file I/O.
+    ///     - reportHeaders: A closure called when the response headers are available.
+    ///     - reportProgress: A closure called when a body chunk has been downloaded, with
+    ///       the total byte count and download byte count passed to it as arguments.
+    public init(
+        path: String,
+        pool: NIOThreadPool = NIOThreadPool(numberOfThreads: 1),
+        reportHeaders: ((HTTPHeaders) -> Void)? = nil,
+        reportProgress: ((_ totalBytes: Int?, _ receivedBytes: Int) -> Void)? = nil
```

Comment:
Can we try to establish certainty about what threading context will be used to invoke these callbacks? It's good to write these things down in documentation.

Reply:
I don't think there's any documentation whatsoever on the threading context of `HTTPClientResponseDelegate` itself, can I even make any assumptions in absence of that?

Reply:
You can say that the callbacks will be invoked in the same threading context that the delegate itself is invoked, as controlled by `EventLoopPreference`. That is covered elsewhere in the docs.

Reply:
Would you mind pointing what this "elsewhere" is? Unfortunately, the Jazzy docs are only searchable for exact symbol names, none of my search queries for threading contexts led to anything.

Reply:
Sure, it's documented on [these fields](https://github.com/swift-server/async-http-client/blob/63fa13ce706f16c9e374f454f59e1c5c26d4598f/Sources/AsyncHTTPClient/HTTPClient.swift#L764-L782).